contents/java/JavaConcurrency.md

```diff
@@ -0,0 +1,149 @@
+
+A thread: 
+* is a lightweight process that exists within a process 
+* is an independent path of execution through program code 
+* has its own call stack and can access shared data of other threads in the same process. 
+
+When multiple threads execute code in parallel of other threads running, it gives rise to multi-threading.
+
+### Creating and Starting Threads
+There are two ways to create a thread in Java:
+ 
+1. Extend the `Thread` class
+1. Implement the `Runnable` interface
+
+After which, override the `run()` method which contains the code that will be executed when the thread starts. 
+
+Depending on how you created the `Thread` you can either create the class that extends the `Thread` class or pass the class that implements the `Runnable` interface into the Thread constructor and then start it.
+
+When deciding which method to use to create a `Thread`, it is always advisable to implement the `Runnable` interface as this results in composition which will allow your code to be more loosely coupled as compared to inheritance. Furthermore, you can extend another class if need be.
+
+### Operations on Threads
+After creating and starting threads, you can carry out operations on them. There are several such operations you can use to manipulate threads:
+
+* Making a thread sleep for a given amount of time. This is useful for pausing the execution of a certain thread for a certain amount of time to free up resources for other threads. The sleeping thread can be interrupted by another thread, which can be a way of communicating that the other thread is done with it's work allowing the current thread to start.  
+```java
+try {
+    Thread.sleep(1000);
+} catch (InterruptedException e){
+    // handle interruption
+}
+```
+
+* Join one thread to the current thread. This will pause the execution of the current thread and cause it to wait until the newly joined thread has finished it's execution. This is useful in cases where there is a need for the results of another thread before moving on.
+```java
+try {
+    thread2.join();
+} catch (InterruptedException e) {
+    // handle interruption
+}
+```
+
+### Thread Interference
+Since threads share the resources of the process they exists in, there will inevitably be conflicts when using shared resources due to the unpredictable nature of threads. When threads try to write to the same resource, thread interference occurs. To illustrate this problem, here's a sequence of execution for two threads, A and B, that increment and decrement a counter variable respectively:
+
+1. Thread A retrieves the value of counter and increments it to a value of 1
+2. Thread B then retrieves the values of counter, which is still 0, and decrements it to a value of -1
+3. Thread A stores the value of 1 to the counter variable
+4. Thread B stores the value of -1 to the counter variable and overwrites the previous value stored by Thread A
```

Comment:
I think you can use images to illustrate this, which will make it easier to understand.